src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +32,131 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Applies changes to the given document so that the folders, name, and path from <paramref name="newDocumentInfo"/>
+        /// are used. Will also perform symbol changes to the code in the document, attempting to sync type and namespace
+        /// to the new information.
+        /// 
+        /// Note that regardless of symbol changes, the document info will be applied and returned in the new solution. 
+        /// </summary>
+        /// <remarks>
+        /// Casing changes in document info will be applied but will not result in any symbol changes.
+        /// </remarks>
+        /// <returns><see cref="RenameDocumentResult"/> with the modified solution and indication of any symbol changes made</returns>
+        public static async Task<RenameDocumentResult> RenameDocumentAsync(
+            Document document, DocumentInfo newDocumentInfo, OptionSet optionSet, CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentInfo.FilePath == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentInfo.FilePath));
+            }
+
+            var newName = newDocumentInfo.Name;
+
+            // Casing changes to files and folders should not trigger any renaming.
+            var shouldRenameType = !string.Equals(document.Name, newName, StringComparison.OrdinalIgnoreCase);
+            var shouldMoveNamespace = !document.Folders.SequenceEqual(newDocumentInfo.Folders, StringComparer.OrdinalIgnoreCase);
+
+            var solution = document.Project.Solution
+                .WithDocumentName(document.Id, newDocumentInfo.Name)
+                .WithDocumentFilePath(document.Id, newDocumentInfo.FilePath)
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
```

Comment:
I'm trying to understand why we need all three -- if you are requiring a name and folders, isn't that sufficient and you don't need the file path? Or the file path has to change in sync? You're kinda caught in an ugly part of the Workspaces API which I don't envy you for, but how these are supposed to interact isn't clear to me.